### PR TITLE
Collapsible panels for the block's toolbar

### LINKED ISSUE #9075: Allow toolbar segments to collapse (responsive toolbar)
When blocks grow thin, whether shown on mobile or in tight nested contexts, the block toolbar currently overflows the boundaries of the block itself. It is not as responsive as it should be.

screen shot 2018-08-17 at 11 10 26

Let's try and mitigate that, by allowing sections of the toolbar to collapse. 

When there's space, show everything:

wide

When space becomes a premium, collapse sections into single icons with dropdowns:

thin

thin w popout

Let's start by doing this for alignments (both decorative and text alignment).

The rules of the block toolbar is that they contain, in order:

- Block Chip (indicator and switcher)
- Decorative alignments
- Text alignments
- Inline formatting
- With https://github.com/WordPress/gutenberg/issues/9074 implemented, More button

However you'd never see both the Decorative alignments and text-alignments/inline formatting together. So if we implemented collapsing for all sections, this would be the minimum size of the toolbar:

screen shot 2018-08-17 at 11 02 53

Note that this uses a new icon for formatting — we'd want to discuss that icon first. But this would be an extreme version of toolbar-section-collapsing.

We could look at employing https://github.com/marcj/css-element-queries to help with this.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #9075

This PR adds 

 - Support for "collapsible" toolbars using the `isCollapsed` prop
 - Support for collapsing text alignment toolbars
 - Auto-collapse text alignment toolbars on mobile
 - Always Collapse text alignment for the paragraph block.

I tried adding auto computation on whether or not to collapse the toolbar segments based on the available space. It turns out it's more challenging than I expected originally and may not be achievable in a not fragile way, I'm instead suggesting it should be "opt-in" like the paragraph block here or more simple computations like "mobile" automatically triggering collapsed panels.

### CODE REVIEW

**Comment by jasmussen** [comment]
This is a very nice first stab. GIF:

toolbar segments

I'd like to do some stule changes — tweak the dropdown arrow a bit, tweak the focus and active styles, that sort of thing. But this is working very well.

It is rather unfortunate that we can't auto-collapse when space is at a premium, and uncollapse when space is aplenty. Is there anything we can do to make this easier? Are there any changes in the design or structure that could accommodate this collapsing, even if those changes come with their own tradeoffs?

Did anything come of the element query discussions? I had my doubts about using polyfills here given how react creates the DOM, but hope springs eternal.

**Comment by youknowriad** [comment]
> Are there any changes in the design or structure that could accommodate this collapsing, even if those changes come with their own tradeoffs?

It's not about design structure but more about Block API and the way we declare the toolbars. 


> Did anything come of the element query discussions?

element query is about the size of the container not about the available space (like media queries) so not really, they can't help a lot

File: packages/editor/src/components/alignment-toolbar/index.js [hunk 1]
```diff
@@ -22,13 +28,18 @@ const ALIGNMENT_CONTROLS = [
...
+function AlignmentToolbar( { isCollapsed, isLargeViewport, value, onChange } ) {
 	function applyOrUnset( align ) {
 		return () => onChange( value === align ? undefined : align );
 	}
 
+	const activeAlignment = find( ALIGNMENT_CONTROLS, ( control ) => control.align === value );
+
 	return (
 		<Toolbar
+			isCollapsed={ isCollapsed || ! isLargeViewport }
```
Comment (jorgefilipecosta) [inline packages/editor/src/components/alignment-toolbar/index.js]: I think we may need an option to disable this forced collapsing on mobile. On the heading block the AlignmentToolbar component is used on the sidebar and collapsing it makes the result feel strange.
sep-07-2018 19-13-29-1

If the resolution is really small it becomes even impossible to change the alignment, that seems unexpected and I'm not certain of the reason why it happens:
sep-07-2018 19-14-03-2

**Review by jorgefilipecosta — COMMENTED** [review]
This is looking good 👍  I only found a problem with headings on mobile because of the way the headings on mobile use the AlignmentToolbar.

**Comment by jorgefilipecosta** [comment]
> It is rather unfortunate that we can't auto-collapse when space is at a premium, and uncollapse when space is aplenty. Is there anything we can do to make this easier?

Maybe we can get this result with some assumptions.  Like on mobile we can consider space a premium like we do and always collapse. On the desktop for the paragraph case, if it is not nested, we can decide that we always have space and never collapse. If it is nested depends on the parent to determine if there is space or not and the parent may decide if the paragraph alignment options are collapsed or not. To pursue this option, we would need to implement some kind of options API that allows parents to control the UI of children as referred in https://github.com/WordPress/gutenberg/issues/7931.
A more simple version would be for the alignment options of the paragraph to be collapsed when the paragraph is nested and uncollapsed when it is not nested (we can get if the paragraph is nested or not with a simple withSelect call). But even the simplest version adds some complexity to a block that is expected to be simple like the paragraph.

**Comment by jasmussen** [comment]
I like this line of thinking, Jorgé. It brings us benefits immediately, despite the challenges. 

Non-mobile, non-nested blocks that can fit the expanded segments should have them. Mobile, nested, or floated blocks, should have collapsed segments. Seems like a great compromise.

**Comment by youknowriad** [comment]
I updated this PR and I'm now auto-collapsing the alignment toolbar on:

 - small screens
 - nested context when the toolbar is not fixed at the top

I didn't add the collapsing when the block is "floated" because the block being floated or not is a block attribute value (the attribute name can change) and can't be detected in a generic way.

**Comment by jasmussen** [comment]
Awesome work Riad, this is working really well. Mobile:

screen shot 2018-09-12 at 10 52 01

Nested:

screen shot 2018-09-12 at 10 51 24

Can we add this for Image/Gallery/Cover Image or in general decorative alignemnts too?

screen shot 2018-09-12 at 10 50 55

I love how the menu is fully tabbable as it should be. 

Let me know when's a good time and I'll push some polish to the indicators — both the dropdown triangle and the focus styles could use a little love, and we need the "active" indicators to show in the dropdown as well — which I can probably fix.

**Comment by youknowriad** [comment]
I added the "collapsible" behavior to the block alignment toolbar as well

**Review by jorgefilipecosta — COMMENTED** [review]
This PR is looking good. I did some tests, and things seem to be working fine.
Maybe we can add some end two end tests for this behaviors so we can catch regressions easily (can be done in a separate PR).
As @jasmussen referred, it would be nice to have indications on the dropdown for the currently active item.
I would say that for example for the block alignment we should also add a gray mark to the icon (even before drop-down is open) if any alignment is enabled (identical to the marks we had before), so the users can differentiate between no alignment and left alignment even without opening the drop-down.
Right now it is impossible to differentiate both states (using the block toolbar):
image

**Comment by youknowriad** [comment]
I added a background on the active item similar to how it shows in the toolbar.

**Comment by jasmussen** [comment]
Thanks for that, Riad. I will be using that to polish it a bit further.

**Comment by jasmussen** [comment]
Okay, I've pushed some polish. Sorry it took so long. 

- Dropdown arrow now the same as in switcher
- Formatting menu items look and behave the same in the dropdown
- Dropdown itself shows the active state, it doesn't show the "toggled" state.

new styles

**Comment by jasmussen** [comment]
Given a sanity check of https://github.com/WordPress/gutenberg/pull/9687/commits/c933243c7adf5100b8460c51fb7c10812308baa6 and https://github.com/WordPress/gutenberg/pull/9687/commits/89790a6158da2bc40dae025b769f69d052b3ff4e, I think this is good to go experience wise. 👍 👍 amazing work Riad!

File: packages/components/src/dropdown-menu/style.scss [hunk 2]
```diff
@@ -78,16 +78,24 @@
...
-		&:hover:not(:disabled):not([aria-disabled="true"]):not(.is-default) {
-			@include formatting-button-style__hover();
+		// Plain menu styles.
+		&:focus:not(:disabled):not([aria-disabled="true"]):not(.is-default) {
+			@include menu-style__focus();
 		}
 
-		&:not(:disabled):not([aria-disabled="true"]):not(.is-default).is-active {
-			@include formatting-button-style__active();
+		// Formatting buttons
```
Comment (youknowriad) [inline packages/components/src/dropdown-menu/style.scss]: It's not only for formatting buttons though it's for any Dropdown menu containing icons right? (like the table block)

Comment (jasmussen) [reply packages/components/src/dropdown-menu/style.scss]: Right, but the table block will never have an is-active applied, so the dark background won't be applied.

File: packages/components/src/dropdown-menu/index.js [hunk 3]
```diff
@@ -64,26 +67,34 @@ function DropdownMenu( {
...
-								disabled={ control.isDisabled }
-							>
-								{ control.title }
-							</IconButton>
+						{ flatMap( controlSets, ( controlSet, indexOfSet ) => (
+							controlSet.map( ( control, indexOfControl ) => (
+								<IconButton
+									key={ [ indexOfSet, indexOfControl ].join() }
+									onClick={ ( event ) => {
+										if ( control.isDisabled ) {
```
Comment (aduth) [inline packages/components/src/dropdown-menu/index.js]: Why should we expect `onClick` callback would ever be triggered when `disabled` prop is passed?

Speaking more from a high-level; this doesn't seem like it should happen.

**Edit:** Observing this is a hold-over from previous implementation. My concern still stands, but not necessarily within scope of the pull request.

File: packages/editor/src/components/alignment-toolbar/index.js [hunk 4]
```diff
@@ -42,3 +60,21 @@ export default function AlignmentToolbar( { value, onChange } ) {
...
+		return {
+			clientId,
+		};
+	} ),
+	withViewportMatch( { isLargeViewport: 'medium' } ),
+	withSelect( ( select, { clientId, isLargeViewport, isCollapsed } ) => {
+		const { getBlockRootClientId, getEditorSettings } = select( 'core/editor' );
+		return {
+			isCollapsed: isCollapsed || ! isLargeViewport || (
+				getBlockRootClientId( clientId ) &&
```
Comment (aduth) [inline packages/editor/src/components/alignment-toolbar/index.js]: Nitpick: `getBlockRootClientId`'s implementation is computationally intensive (aside: it should be improved), and thus we could reorder the `&&` parts here to short-circuit on `getEditorSettings` as the fast path.

**Review by aduth — APPROVED** [review]

**Comment by mtias** [comment]
Nice to see this in.

**Pull request merged**